File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/list-ops/list_ops.py
def append(list1, list2):
    return list1 + list2


def concat(lists):
    result = []
    for lst in lists:
        result.extend(lst)
    return result


def filter(function, list):
    return [item for item in list if function(item)]


def length(list):
    count = 0
    for _ in list:
        count += 1
    return count


def map(function, list):
    return [function(item) for item in list]


def foldl(function, list, initial):
    accumulator = initial
    for item in list:
        accumulator = function(accumulator, item)
    return accumulator


def foldr(function, list, initial):
    if not list:
        return initial
    return function(list[0], foldr(function, list[1:], initial))


def reverse(list):
    return list[::-1]

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/list-ops/list-ops.js
export class List {
  constructor(values = []) {
    this.values = values;
  }

  append(otherList) {
    const newValues = [...this.values];
    for (const value of otherList.values) {
      newValues.push(value);
    }
    return new List(newValues);
  }

  concat(lists) {
    let newValues = [...this.values];
    for (const list of lists.values) {
      for (const value of list.values) {
        newValues.push(value);
      }
    }
    return new List(newValues);
  }

  filter(predicate) {
    const newValues = [];
    for (const value of this.values) {
      if (predicate(value)) {
        newValues.push(value);
      }
    }
    return new List(newValues);
  }

  map(fn) {
    const newValues = [];
    for (const value of this.values) {
      newValues.push(fn(value));
    }
    return new List(newValues);
  }

  length() {
    let count = 0;
    for (const _ of this.values) {
      count++;
    }
    return count;
  }

  foldl(fn, initial) {
    let accumulator = initial;
    for (const value of this.values) {
      accumulator = fn(accumulator, value);
    }
    return accumulator;
  }

  foldr(fn, initial) {
    let accumulator = initial;
    for (let i = this.values.length - 1; i >= 0; i--) {
      accumulator = fn(accumulator, this.values[i]);
    }
    return accumulator;
  }

  reverse() {
    const newValues = [];
    for (let i = this.values.length - 1; i >= 0; i--) {
      newValues.push(this.values[i]);
    }
    return new List(newValues);
  }
}
